### PR TITLE
Assertion `old_part_id == m_last_part' failed in ha_partition::update…

### PR DESCRIPTION
**Pull request by FooBarrior**
…_row or `part_id == m_last_part' in ha_partition::delete_row upon UPDATE/DELETE after dropping versioning

PRIMARY KEY change hadn't been treated as partition reorganization in case of partitioning by KEY() (without parameters).

* set `*partition_changed= true` in the described case.
* since add/drop system versioning does not affect alter_info->key_list, it required separate attention

### CODE REVIEW

File: sql/sql_partition.cc [hunk 1]
```diff
@@ -5929,6 +5929,23 @@ the generated partition syntax in a correct manner.
...
+      {
+        if (alter_info->flags & (ALTER_DROP_SYSTEM_VERSIONING | ALTER_ADD_SYSTEM_VERSIONING))
+          *partition_changed= true;
+
+        List_iterator<Key> it(alter_info->key_list);
+        Key *key;
+        while((key= it++) || !*partition_changed)
+        {
+          if (key->type == Key::PRIMARY)
+            *partition_changed= true;
```
Comment (midenok) [inline sql/sql_partition.cc]: Add `break;`, remove wrong condition `|| !*partition_changed`.

File: sql/sql_partition.cc [hunk 2]
```diff
@@ -5929,6 +5929,23 @@ the generated partition syntax in a correct manner.
...
       }
+
+      // In case of PARTITION BY KEY(), check if primary key has changed
+      // System versioning also implicitly adds/removes primary key parts
+      if (alter_info->partition_flags == 0 && part_info->list_of_part_fields
+          && part_info->part_field_list.elements == 0)
+      {
+        if (alter_info->flags & (ALTER_DROP_SYSTEM_VERSIONING | ALTER_ADD_SYSTEM_VERSIONING))
+          *partition_changed= true;
+
```
Comment (midenok) [inline sql/sql_partition.cc]: add `else if (!*partition_changed)`

File: sql/sql_partition.cc (hunk 1 above)
Comment (FooBarrior) [reply sql/sql_partition.cc]: or just make it `&&`, right?

Comment (midenok) [reply sql/sql_partition.cc]: No. I'm asking to do more optimal code.

Comment (FooBarrior) [reply sql/sql_partition.cc]: did it

File: sql/sql_partition.cc (hunk 2 above)
Comment (FooBarrior) [reply sql/sql_partition.cc]: did it not exactly that way; see the code

Comment (midenok) [reply sql/sql_partition.cc]: Is it fixed?

File: sql/sql_partition.cc [hunk 3]
```diff
@@ -5929,6 +5929,32 @@ the generated partition syntax in a correct manner.
           *partition_changed= true;
         }
       }
+
+      // In case of PARTITION BY KEY(), check if primary key has changed
+      // System versioning also implicitly adds/removes primary key parts
+      if (alter_info->partition_flags == 0 && part_info->list_of_part_fields
+          && part_info->part_field_list.elements == 0 && !*partition_changed)
```
Comment (midenok) [inline sql/sql_partition.cc]: SQL layer coding style recommends operators at the end of line contrary to InnoDB which recommends operator at the beginning of line. Please see code around if you have doubts how to format.

**Review by midenok — APPROVED** [review]

File: sql/sql_partition.cc (hunk 3 above)
Comment (FooBarrior) [reply sql/sql_partition.cc]: where did you get that??

Comment (FooBarrior) [reply sql/sql_partition.cc]: well, ok, i think you're right

**Pull request merged**